features/text/0001-multiple_service_endpoints.md

```diff
@@ -0,0 +1,323 @@
+
+For example on the host "epic.grnet.gr" there are 2 running instances of Service Type "b2safe.handle.api". Each of these instances is bind on different ports and serves a different set of prefixes. In order to be able to monitor these two instances of the same Service Type, running on the same host, we need to be able to properly model them in the GOCDB and to modify the ARGO NCG component to be able to generate the proper Nagios configuration.
+
+The approach described in this document relies on functionality that is already available in the GOCDB, but requires us to extend the information that we record in the GOCDB and to perform the appropriate modifications to the ARGO NCG component that is responsible for the generation of the Nagios configuration.
+
+
+# Detailed design
+[design]: #detailed-design
+
+### Use case: Support of multiple service types (even same ones) per endpoint with different attributes, extensions and service groups. E.g:
+
+```
+service endpoint: host => epic.grnet.gr/service_type => b2handle.handle.resolver , URL => https://epic.grnet.gr , service_group => B2HANDLE
+
+service endpoint: host => epic.grnet.gr/service_type => b2handle.handle.api , URL => https://epic.grnet.gr/api/v2/handles/foo , service_group => B2HANDLE
+
+service endpoint: host => epic.grnet.gr/service_type => b2handle.handle.api , URL => https://epic.grnet.gr/api/v2/handles/ , service_group => B2HANDLE
+
+service endpoint: host => epic.rzg.mpg.de/service_type => b2handle.handle.api , URL => https://epic.rzg.mpg.de/epic_v2/ , service_group => B2HANDLE
+```
+### Goal: Use a dedicated service endpoint entry on GOCDB for each case.
+
+Separation of attributes, extensions, groups per endpoint using the primary_key of each entry. To retrieve the primary key we parse the XML response from the GOCDB API and store the value of the XML element `PRIMARY_KEY`.
+
+## GOCDB
+In EUDAT GOCDB we use the following model:
+* Service Endpoints are the combination of a hostname and a Service Type
+* Service Endpoints can be grouped together in Service Groups to model an EUDAT Service that is provided to a specific community (e.g. all the Service Points of the B2Handle EUDAT Service that have been assigned to Community X)
+
+The solution proposed is to remove the artificial restriction to have unique Service Endpoints (unique tuples of Service Endpoint and hostname) and to utilise the GOCDB extensions to include richer information where needed. From the point of view of the monitoring service a resource will now be identified as the triplet: Hostname, Service Type, URL. This solution uses existing GOCDB features and requires only the input of extra data in the GOCDB.
+
+Below follows a typical example of the XML structure and the particular information for a service endpoint entry in GOCDB. Notice that in case of epic.grnet.gr there are two separate GOCDB entries with different IDs in order to be possible for us to retrieve the appropriate URL for each case.
```

Comment:
This is not a *typical example*. This is how we propose to model resources in the GOCDB